src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/PatternVariableNameCheck.java

```diff
@@ -0,0 +1,140 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.naming;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks that pattern variable names conform to a specified pattern.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code format} - Specifies valid identifiers. Type is {@code java.util.regex.Pattern}.
+ * Default value is {@code "^[a-z][a-zA-Z0-9]*$"}.
+ * </li>
+ * </ul>
+ * <p>
+ * An example of how to configure the check is:
+ * </p>
+ * <pre>
+ * &lt;module name="PatternVariableName"/&gt;
+ * </pre>
+ * <p>Code Example:</p>
+ * <pre>
+ * class MyClass {
+ *     {
+ *         Object o1 = null;
+ *         if (o1 instanceof String STRING) { // violation, name 'STRING' must
+ *                                            // match pattern '^[a-z][a-zA-Z0-9]*$'
+ *         }
+ *         if (o1 instanceof Integer num) { // OK
+ *         }
+ *     }
+ * }
+ * </pre>
+ * <p>
+ * An example of how to configure the check for names that begin with a lower
+ * case letter, followed by letters, digits, and underscores is:
```

Comment:
Actually, this is a pattern for "a lower case letter, followed by letters and digits, optionally separated by underscore".
These variable names: `foo_`, `foo___bar` do not match the pattern, although they are "begin with a lower
 case letter, followed by letters, digits, and underscores". Please update either the pattern or the description.

Reply:
Thanks for pointing this out, regex is definitely something I need to study more.

Reply:
Fixed.